modules/server_new/src/main/scala/observe/server/engine/Engine.scala

```diff
@@ -137,96 +138,130 @@ class Engine[F[_]: MonadThrow: Logger] private (
-                    )
-                  ) *> send(Event.modifyState(atomLoad(this, id)))
-                // Step execution completed. Check requested stop and breakpoint here.
-                case Some(qs)                              =>
-                  EngineHandle.replaceSequenceState(id)(qs) *>
-                    (if (
-                       qs.getCurrentBreakpoint && !qs.current.execution.exists(_.uninterruptible)
-                     ) {
-                       switch(id)(SequenceState.Idle) *> send(Event.breakpointReached(id))
-                     } else if (seq.isLastAction) {
-                       // Only process stop states after the last action of the step.
-                       if (userStop || internalStop) {
-                         if (qs.current.execution.exists(_.uninterruptible))
-                           send(Event.executing(id)) *> send(Event.stepComplete(id))
-                         else
-                           switch(id)(SequenceState.Idle) *> send(Event.sequencePaused(id))
-                       } else {
-                         // after the last action of the step, we need to reload the sequence
-                         switch(id)(
-                           SequenceState.Running(
-                             userStop,
-                             internalStop,
-                             waitingUserPrompt = false,
-                             waitingNextAtom = true,
-                             starting = false
-                           )
-                         ) *> send(
-                           Event.modifyState(atomReload(this, id, OnAtomReloadAction.StartNewAtom))
-                         )
-                           *> send(Event.stepComplete(id))
-                       }
-                     } else send(Event.executing(id)))
+      .getSequenceState(obsId)
+      .flatMap(seqState =>
+        EngineHandle
+          .debug(
+            s"******* {next} CHECKING NEXT FOR COMPLETION, $obsId, NEXT: ${pprint(seqState.map(_.next))}"
+          ) >>
+          seqState
+            .map { seq =>
+              seq.status match {
+                case SequenceState.Running(userStop, internalStop, _, _, _) =>
+                  seq.next match {
+                    // Empty state
+                    case None                                  =>
+                      EngineHandle.debug(s"**** {next} STOPPING SEQUENCE BECAUSE NEXT IS EMPTY") >>
+                        send(Event.finished(obsId))
+                    // Final State
+                    case Some(qs @ Sequence.State.Final(_, _)) =>
+                      EngineHandle.replaceSequenceState(obsId)(qs) *> switch(obsId)(
+                        SequenceState.Running(
+                          userStop,
+                          internalStop,
+                          waitingUserPrompt = IsWaitingUserPrompt.Yes,
+                          waitingNextAtom = IsWaitingNextAtom.Yes, // TODO Should this be No?
```

Comment:
Here the current state of the observation is `seq`, and we are checking its future state if we were to progress the execution, `seq.next` (the executin state will not progress until we update the engine state). The fact the `seq.next` is `Final` here means that `seq` doesn't have anything else to execute. Therefore, we need to check  in the ODB if there is something new to execute next, or to declare the atom completed.

Reply:
I see, thank you!

I removed the comment.

Anyway, in the future we should remove `waitingNextAtom` and just use the new latch. Otherwise, we have 2 mechanisms for the same goal.